### DISCUSSION TITLE
Some problems about NEST to GeNN.

### DISCUSSION THREAD

**Discussion by ChenLumen** [genn-team/genn#532] [GeNN questions]
Hi, recently I wanted to simulate my model over and over again with GeNN. But I didn't found about how to reset the network. For example, in NEST we can use nest.ResetNetwork() and nest.SetKernel Status() to reset network. And someone can tell me how to achieve this in GeNN? Thanks!

**Comment by neworderofjamie** [comment]
The ``reinitialise()`` method of ``GeNNModel`` does just this

**Reply by ChenLumen** [reply]
Oh, thanks! And I want to add spikes input to my model. Like NEST, I can use it:
`spike_generators = nest.Create('souje_generator', 2)
...
nest.Connect(tuple(range(spike_generators[0], spike_generators[-1]+1)),
                      tuple(range(gids[0], gids[-1]+1)),
                      conn_spec, syn_spec)`
In this way, I created and connected spike generators with my model. And then, I just default the spike times to spike_generators, like:
`nest.SetStatus([spike_generators[0]], {'spike_times': np.array([0., 1., 3.])})
nest.SetStatus([spike_generators[1]], {'spike_times': np.array([0., 2., 5.])})`
To implement this method, I looked for your user manual, but I could only find examples of adding one generator and one soike point. I tried to expand it. But building the model took all night and in the end it gave me an error.
So, does GeNN offer similar functionality in NEST? And another question, if GeNN can do it, can we not rebuild the model? You know it takes a long time to build a model.
Thanks for your help!

**Comment by neworderofjamie** [comment]
Ahh yes there is a built in ``SpikeSourceArray`` model but it's not very well documented. For example, if you want to build a spike source array population with two neurons, the first of which emits spikes in the 10th and 20th timestep and the second which emits spikes in the 5th and 15th you might do this:

```python
spike_times = [[10, 20], [5, 15]]

# Count spikes each neuron should emit
spike_counts = [len(n) for n in spike_times ]

# Get start and end indices of each spike sources section
end_spike = np.cumsum(spike_counts)
start_spike = np.empty_like(end_spike)
start_spike[0] = 0
start_spike[1:] = end_spike[0:-1]

....

ssa = model.add_neuron_population("SSA", 2, "SpikeSourceArray", {}, 
                                  {"startSpike": start_spike, "endSpike": end_spike})

# Set the spike times
ssa.set_extra_global_param("spikeTimes", np.concatenate(spike_times))
```

**Reply by ChenLumen** [reply]
Thanks, and the use this command to add the spike point?
`ssa.set_extra_global_param("spikeTimes", spike_times)`

**Reply by ChenLumen** [reply]
When I used 11596 neurons and 51 spike points, gave me an error:
ValueError: could not broadcast input array from shape (11596, 51) into shape (11596)

**Reply by neworderofjamie** [reply]
Apologies, forgot that bit - have updated the code. The only bit you're missing is that you need to concatenate the per-neuron spike times together

**Reply by ChenLumen** [reply]
OK, thanks. However, I encounted another problem. When the number of neuron spikes was increasing, the model simulation speed became very slow. I used 1156 neurons, the maximum number of spike emitted was 51, the matrix_type="PROCEUDRAL_PROCEDURALG". I simulated the time of 1650ms with a time step of 1ms. But 3 hours after model.build() and model.load() passed, the work was still not finished. It's worth noting that when I changed the simulation time to 700ms, it took just a few seconds. Do you know the reason? Thank you very much!

**Reply by ChenLumen** [reply]
And there were 66,269 spikes. Is it because there area too many spikes to add that GeNN does not support?

**Reply by neworderofjamie** [reply]
Strange....few questions:
- What exactly is the model you're simulating and can you share it?
- What backend are you using (```print(model.use_backend)``` will give you this information)?
- Is there a particular reason you're using ``PROCEDURAL_PROCEDURALG``? In general, this is only used when the model is too large to fit in GPU memory,
